File: benchmarks2d/heightfield2.rs
use rapier_testbed2d::Testbed;
use rapier2d::na::DVector;
use rapier2d::prelude::*;

pub fn init_world(testbed: &mut Testbed) {
    /*
     * World
     */
    let mut bodies = RigidBodySet::new();
    let mut colliders = ColliderSet::new();
    let impulse_joints = ImpulseJointSet::new();
    let multibody_joints = MultibodyJointSet::new();

    /*
     * Ground
     */
    let ground_size = Vector::new(50.0, 1.0);
    let nsubdivs = 2000;

    let heights = DVector::from_fn(nsubdivs + 1, |i, _| {
        if i == 0 || i == nsubdivs {
            80.0
        } else {
            (i as f32 * ground_size.x / (nsubdivs as f32)).cos() * 2.0
        }
    });

    let rigid_body = RigidBodyBuilder::fixed();
    let handle = bodies.insert(rigid_body);
    let collider = ColliderBuilder::heightfield(heights, ground_size);
    colliders.insert_with_parent(collider, handle, &mut bodies);

    /*
     * Create the cubes
     */
    let num = 26;
    let rad = 0.5;

    let shift = rad * 2.0;
    let centerx = shift * (num / 2) as f32;
    let centery = shift / 2.0;

    for i in 0..num {
        for j in 0usize..num * 5 {
            let x = i as f32 * shift - centerx;
            let y = j as f32 * shift + centery + 3.0;

            // Build the rigid body.
            let rigid_body = RigidBodyBuilder::dynamic().translation(vector![x, y]);
            let handle = bodies.insert(rigid_body);

            if j % 2 == 0 {
                let collider = ColliderBuilder::cuboid(rad, rad);
                colliders.insert_with_parent(collider, handle, &mut bodies);
            } else {
                let collider = ColliderBuilder::ball(rad);
                colliders.insert_with_parent(collider, handle, &mut bodies);
            }
        }
    }

    /*
     * Set up the testbed.
     */
    testbed.set_world(bodies, colliders, impulse_joints, multibody_joints);
    testbed.look_at(point![0.0, 50.0], 10.0);
}

File: python/rapier-testbed/rapier_testbed/examples2/heightfield2.py
"""Port of examples2d/heightfield2.rs."""
from __future__ import annotations

import math

import numpy as np

import rapier2d as rp
from .._registry import register

CATEGORY = "Collisions"
NAME = "Heightfield"


def init_world(testbed) -> None:
    bodies = rp.RigidBodySet()
    colliders = rp.ColliderSet()
    impulse_joints = rp.ImpulseJointSet()
    multibody_joints = rp.MultibodyJointSet()

    ground_size = (50.0, 1.0)
    nsubdivs = 2000

    heights = np.zeros(nsubdivs + 1, dtype=np.float32)
    for i in range(nsubdivs + 1):
        if i == 0 or i == nsubdivs:
            heights[i] = 8.0
        else:
            heights[i] = math.cos(i * ground_size[0] / nsubdivs) * 2.0

    handle = bodies.insert(rp.RigidBody.fixed())
    colliders.insert_with_parent(
        rp.Collider.heightfield(heights, ground_size), handle, bodies
    )

    num = 20
    rad = 0.5
    shift = rad * 2.0
    centerx = shift * (num // 2)
    centery = shift / 2.0

    for i in range(num):
        for j in range(num):
            x = i * shift - centerx
            y = j * shift + centery + 3.0
            handle = bodies.insert(rp.RigidBody.dynamic().translation((x, y)))
            if j % 2 == 0:
                colliders.insert_with_parent(
                    rp.Collider.cuboid(rad, rad), handle, bodies
                )
            else:
                colliders.insert_with_parent(rp.Collider.ball(rad), handle, bodies)

    testbed.set_world(bodies, colliders, impulse_joints, multibody_joints)
    testbed.set_camera_2d(center=(0.0, 0.0), zoom=10.0)


register(CATEGORY, NAME, init_world, dim=2)

if __name__ == "__main__":
    from .._picker import run

    run(initial=f"{CATEGORY} / {NAME}")
